hca/dss/__init__.py

```diff
@@ -38,24 +42,86 @@ def download(self, bundle_uuid, replica, version="", dest_name=""):
-                    while True:
-                        chunk = fh.raw.read(1024*1024)
-                        if len(chunk) == 0:
+            # Attempt to download the data.  If a retryable exception occurs, we wait a bit and retry again.  The delay
+            # increases each time we fail and decreases each time we successfully read a block.  We set a quota for the
+            # number of failures that goes up with every successful block read and down with each failure.
+            #
+            # If we can, we will attempt HTTP resume.  However, we verify that the server supports HTTP resume.  If the
+            # ranged get doesn't yield the correct header, then we start over.
+            delay = min_delay_seconds
+            retries_left = initial_retries_left
+            with open(file_path, "wb") as fh, ChecksummingSink() as checksummer:
+                while True:
+                    try:
+                        response = self.get_file._request(
+                            dict(uuid=file_uuid, replica=replica),
+                            stream=True,
+                            headers={
+                                'Range': "bytes={}-".format(fh.tell())
+                            },
+                        )
+                        try:
+                            if not response.ok:
+                                logger.error("%s", "File {}: GET FAILED.".format(filename))
+                                logger.error("%s", "Response: {}".format(response.text))
+                                break
+
+                            consume_bytes = int(fh.tell())
+                            server_start = 0
+                            content_range_header = response.headers.get('Content-Range', None)
+                            if content_range_header is not None:
+                                cre = re.compile("bytes (\d+)-(\d+)")
+                                mo = cre.search(content_range_header)
+                                if mo is not None:
+                                    server_start = int(mo.group(1))
+
+                            consume_bytes -= server_start
+                            assert consume_bytes >= 0
+                            if server_start > 0 and consume_bytes == 0:
+                                logger.info("%s", "File {}: Resuming at {}.".format(
+                                    filename, server_start))
+                            elif consume_bytes > 0:
+                                logger.info("%s", "File {}: Resuming at {}. Dropping {} bytes to match".format(
+                                    filename, server_start, consume_bytes))
+
+                                while consume_bytes > 0:
+                                    bytes_to_read = min(consume_bytes, 1024*1024)
+                                    content = response.iter_content(chunk_size=bytes_to_read)
```

Comment:
Is there a reason why you use response.iter_content instead of response.raw.read?

Reply:
It's a documented API.  `response.raw.read` is not.

Also, the implementation of response.iter_content seems to branch on whether the backend is urllib3 or not, and in the case of urllib3, it doesn't even call `response.raw.read`.

Reply:
Hmm ok. I think it is documented (http://docs.python-requests.org/en/master/api/#requests.Response.raw and http://urllib3.readthedocs.io/en/latest/reference/index.html#urllib3.response.HTTPResponse.read) and I'm vaguely unhappy with the extra indirection happening here for no discernible reason, but I guess it's OK.

Reply:
So as an example, `iter_content(..)` in requests calls `raw.stream(decode_content=True)` for urllib3.  If I just call `read(..)`, it will not call with `decode_content=True`.